packages/shell-api/src/shell-internal-state.ts

```diff
@@ -0,0 +1,238 @@
+import {
+  AggregationCursor,
+  CursorIterationResult,
+  Cursor,
+  Help,
+  Database,
+  Mongo,
+  ReplicaSet,
+  Shard,
+  signatures,
+  ShellBson,
+  toIterator,
+} from './index';
+import { EventEmitter } from 'events';
+import { DatabaseOptions, Document, ServiceProvider, ReplPlatform } from '@mongosh/service-provider-core';
+import { MongoshInvalidInputError, MongoshUnimplementedError } from '@mongosh/errors';
+import AsyncWriter from '@mongosh/async-rewriter';
+
+/**
+ * Anything to do with the internal shell state is stored here.
+ */
+export default class ShellInternalState {
+  public currentCursor: Cursor | AggregationCursor;
+  public currentDb: Database;
+  public messageBus: EventEmitter;
+  public asyncWriter: AsyncWriter;
+  public initialServiceProvider: ServiceProvider; // the initial service provider
+  public uri: string;
+  public connectionInfo: any;
+  constructor(initialServiceProvider: ServiceProvider, messageBus: any = new EventEmitter()) {
+    this.initialServiceProvider = initialServiceProvider;
+    this.messageBus = messageBus;
+    this.asyncWriter = new AsyncWriter(signatures);
+    const mongo = new Mongo(this);
+    this.currentCursor = null;
+    this.currentDb = mongo.getDB('test'); // TODO: set to CLI arg
+  }
+
+  async fetchConnectionInfo(): Promise<void> {
+    this.connectionInfo = await this.currentDb.mongo.serviceProvider.getConnectionInfo();
+    this.messageBus.emit('mongosh:connect', this.connectionInfo.extraInfo);
+  }
+
+  // TODO: do we need to close all Mongos ever made?
+  close(p): void {
+    try {
+      this.currentDb.mongo.serviceProvider.close(p);
```

Comment:
`serviceProvider .close` returns a promise, this `try` `catch` would not defend us much from an unhandled rejection. Probably `await` or `.serviceProvider.close(p).catch(() => {})`.

Reply:
Good catch! The try/catch is there in case `currentDb` isn't defined (if the user resets `db` to something strange) but you're totally right that we should also check for errors thrown by actually closing